File: Cpp-C/Ema/Src/Access/Impl/ServiceEndpointDiscovery.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2019-2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

#include "ServiceEndpointDiscovery.h"
#include "ServiceEndpointDiscoveryClient.h"
#include "ServiceEndpointDiscoveryOption.h"
#include "ServiceEndpointDiscoveryImpl.h"
#include "ServiceEndpointDiscoveryConfig.h"
#include "ExceptionTranslator.h"

using namespace refinitiv::ema::access;

ServiceEndpointDiscovery::ServiceEndpointDiscovery()
	: _pImpl(0)
{
	try
	{
		_pImpl = new ServiceEndpointDiscoveryImpl(this, NULL);
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory for ServiceEndpointDiscoveryImpl in ServiceEndpointDiscovery().");
	}
}

ServiceEndpointDiscovery::ServiceEndpointDiscovery(const ServiceEndpointDiscoveryConfig& serviceEndpointDiscoveryConfig)
	: _pImpl(0)
{
	try
	{
		_pImpl = new ServiceEndpointDiscoveryImpl(this, &serviceEndpointDiscoveryConfig);
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory for ServiceEndpointDiscoveryImpl in ServiceEndpointDiscovery( const ServiceEndpointDiscoveryConfig& ).");
	}
}

ServiceEndpointDiscovery::ServiceEndpointDiscovery(const EmaString& tokenServiceURLV1)
	: _pImpl(0)
{
	ServiceEndpointDiscoveryConfig config;
	config.tokenServiceUrlV1(tokenServiceURLV1);

	try
	{
		_pImpl = new ServiceEndpointDiscoveryImpl(this, &config);
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory for ServiceEndpointDiscoveryImpl in ServiceEndpointDiscovery( const EmaString& ).");
	}
}

ServiceEndpointDiscovery::ServiceEndpointDiscovery(const EmaString& tokenServiceURLV1, const EmaString& serviceDiscoveryURL)
: _pImpl(0)
{
	ServiceEndpointDiscoveryConfig config;
	config.tokenServiceUrlV1(tokenServiceURLV1);
	config.serviceDiscoveryUrl(serviceDiscoveryURL);

	try
	{
		_pImpl = new ServiceEndpointDiscoveryImpl(this, &config);
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory for ServiceEndpointDiscoveryImpl in ServiceEndpointDiscovery( const EmaString&, const EmaString& ).");
	}
}

ServiceEndpointDiscovery::ServiceEndpointDiscovery(const EmaString& tokenServiceURLV1, const EmaString& tokenServiceURLV2, const EmaString& serviceDiscoveryURL)
	: _pImpl(0)
{
	ServiceEndpointDiscoveryConfig config;
	config.tokenServiceUrlV1(tokenServiceURLV1);
	config.tokenServiceUrlV2(tokenServiceURLV2);
	config.serviceDiscoveryUrl(serviceDiscoveryURL);


	try
	{
		_pImpl = new ServiceEndpointDiscoveryImpl(this, &config);
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory for ServiceEndpointDiscoveryImpl in ServiceEndpointDiscovery( const EmaString&, const EmaString&, const EmaString& ).");
	}
}

ServiceEndpointDiscovery::~ServiceEndpointDiscovery()
{
	if (_pImpl)
	{
		delete _pImpl;
		_pImpl = 0;
	}
}

void ServiceEndpointDiscovery::registerClient(const ServiceEndpointDiscoveryOption& params, ServiceEndpointDiscoveryClient& client,void *closure)
{
	_pImpl->registerClient(params, client, closure);
}

File: Java/Ema/Core/src/main/java/com/refinitiv/ema/access/ServiceEndpointDiscovery.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
// *|                See the project's LICENSE.md for details.
// *|           Copyright (C) 2019 LSEG. All rights reserved.     
///*|-----------------------------------------------------------------------------

package com.refinitiv.ema.access;

/**
 * ServiceEndpointDiscovery class provides the functionality to query endpoints from LDP service discovery.
 *
 * <p>Application interacts with the service discovery through the ServiceEndpointDiscovery interface methods.<br>
 * The results of these interactions are communicated back to application through ServiceEndpointDiscoveryClient.</p>
 * 
 * <p>An ServiceEndpointDiscovery is created from EmaFactory<br>
 * (see {@link com.refinitiv.ema.access.EmaFactory#createServiceEndpointDiscovery()}
 *  or {@link com.refinitiv.ema.access.EmaFactory#createServiceEndpointDiscovery(String, String)}).</p>
 *
 * The following code snippet shows basic usage of ServiceEndpointDiscovery class in a simple consumer type app.
 *
 * <pre>
 * // create an implementation for ServiceEndpointDiscoveryClient to process service discovery responses
 * class AppClient implements ServiceEndpointDiscoveryClient
 * {
 * 		void onSuccess(ServiceEndpointDiscoveryResp serviceEndpointDiscoveryResp, ServiceEndpointDiscoveryEvent event);
 *		void onError(String errorText, ServiceEndpointDiscoveryEvent event);
 * }
 * 
 * public class Consumer
 * {
 *		public static void main(String[] args)
 *		{
 *			ServiceEndpointDiscovery serviceEndpointDiscovery = null;
 *			try
 *			{
 *				AppClient appClient = new AppClient();
 *
 *				// instantiate ServiceEndpointDiscovery object
 *				serviceEndpointDiscovery = EmaFactory.createServiceEndpointDiscovery();
 *
 *				// Query endpoints from LDP service discovery
 *				serviceEndpointDiscovery.registerClient(EmaFactory.createServiceEndpointDiscoveryOpion().username(userName)
 *							.password(password).transport(ServiceEndpointDiscoveryOption::TCP).clientId(clientId), appClient);
 *			}
 *			catch (OmmException excp)
 *       	{
 *          	System.out.println(excp.getMessage());
 *       	}
 *       	finally 
 *       	{
 *          	if (serviceEndpointDiscovery != null) serviceEndpointDiscovery.uninitialize();
 *       	}
 *		}
 * }
 * </pre>
 * 
 * @see ServiceEndpointDiscoveryOption
 * @see ServiceEndpointDiscoveryClient
 * @see OmmException
 */
public interface ServiceEndpointDiscovery
{
	/** 
	 * Queries the LDP service discovery synchronously to get endpoints according to the specified parameters
	 * 
	 * <p>This method is ObjectLevelSafe.</p>
	 * 
	 * @throws OmmInvalidUsageException if application passes invalid ServiceEndpointDiscoveryOption
	 * 
	 * @param params specifies query options to get endpoints
	 * @param client specifies ServiceEndpointDiscoveryClient instance receiving notifications about this query
	 */
	public void registerClient(ServiceEndpointDiscoveryOption params, ServiceEndpointDiscoveryClient client);
	
	/** 
	 * Queries the LDP service discovery synchronously to get endpoints according to the specified parameters
	 * 
	 * <p>This method is ObjectLevelSafe.</p>
	 * 
	 * @throws OmmInvalidUsageException if application passes invalid ServiceEndpointDiscoveryOption
	 * 
	 * @param params specifies query options to get endpoints
	 * @param client specifies ServiceEndpointDiscoveryClient instance receiving notifications about this query
	 * @param closure specifies application defined query identification
	 */
	public void registerClient(ServiceEndpointDiscoveryOption params, ServiceEndpointDiscoveryClient client, Object closure);
	
	/**
	 * Uninitializes the ServiceEndpointDiscovery object.
	 * 
	 * <p> This method is ObjectLevelSafe.</p>
	 */
	public void uninitialize();
}
